### DISCUSSION TITLE
Can the Read MagickImageCollection method be faster?

### DISCUSSION THREAD

**Discussion by manuel-collab** [dlemstra/Magick.NET#1710] [Help]
Hello Dirk,

First of all, thank you for the great effort involved in maintaining these bookstores that help the community.

I'm fairly new to implementing this library and would need help on increasing the imaging performance of a document. Are threads used internally to extract them?

How could the images be extracted in parallel effectively without having to read the document completely?

Thank you very much for your help.

Best regards

**Comment by dlemstra** [comment]
You are reading a document?

**Comment by manuel-collab** [comment]
Yes, I am reading the pdf document, which is also about 200 pages long. How could rendering these pages be implemented faster? The minimum DPI I need to use is 150

**Reply by dlemstra** [reply]
With 200 pages your are might be using a lot of memory. You could speed this up by reading it page per page. With `PdfInfo.Create(` you can get the page count and then do a for loop and only read that scene.

**Comment by manuel-collab** [comment]
I attach a large file with which I am testing. How could you get good image extraction time?

Thank you
[document.pdf](https://github.com/user-attachments/files/16923171/document.pdf)

**Comment by manuel-collab** [comment]
This is the code I use, it takes about 5 minutes could it be improved in time?:

[Fact]
public void Test1()
{
    ResourceLimits.LimitMemory(new Percentage(5));

    bool result = false;
    int totalPages = 0;

    try
    {
        MagickImageCollection magickImages = new MagickImageCollection();

        MagickReadSettings settings1 = new MagickReadSettings()
        {
            ColorSpace = ColorSpace.sRGB,
            ColorType = ColorType.TrueColor,
            Compression = CompressionMethod.JPEG,
            Density = new Density(300, 300)
        };

        magickImages.Read(@"C:\temp\document.pdf", settings1);

        Parallel.ForEach(magickImages, new ParallelOptions { MaxDegreeOfParallelism = System.Environment.ProcessorCount }, (image, state, index) =>
        {
            try
            {
                image.Quality = 50;
                image.Format = MagickFormat.Jpg;
                image.Alpha(AlphaOption.Remove);

                image.Write(System.IO.Path.Combine(@"C:\temp\Pruebas\", (index + 1) + ".jpg"));

            }
            catch (Exception)
            {
            }
            finally 
            {
                if (image.IsDisposed == false)
                {
                    image.Dispose();
                }
            }
        });

        magickImages.Dispose();

        result = true;
    }
    catch (Exception)
    {
        throw;
    }

    Assert.True(result);
}

**Comment by manuel-collab** [comment]
With FrameIndex from MagickReadSettings?

Un saludo,

Manuel Quero Delgado

**Comment by dlemstra** [comment]
You should not use the `MagickImageCollection` but read a `MagickImage` and specify the `FrameIndex` in the `MagickReadSettings` inside a `Parallel.For` loop that use the information that was acquired with `PdfInfo`.

**Comment by manuel-collab** [comment]
Thanks Kirk, but how can you automatically delete the temporary files to
generate the reading of each of the images? magick-*

El lun, 9 sept 2024 a las 21:15, Dirk Lemstra ***@***.***>)
escribió:

> You should not use the MagickImageCollection but read a MagickImage and
> specify the FrameIndex in the MagickReadSettings inside a Parallel.For
> loop that use the information that was acquired with PdfInfo.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dlemstra/Magick.NET/discussions/1710#discussioncomment-10594872>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APJ2PEWWQOUVVOGHQ3YSWZDZVXXV5AVCNFSM6AAAAABN25F25SVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJZGQ4DOMQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Un saludo,

Manuel Quero Delgado

**Comment by dlemstra** [comment]
It sounds like you are missing a `using` and not disposing an image correctly.

**Reply by manuel-collab** [reply]
In a Parallel.For loop, how do I successfully release MagickImage instances?

**Comment by manuel-collab** [comment]
This is my source code:

[Fact]
public void GetImageFrameToFrame()
{
    ResourceLimits.LimitMemory(new Percentage(5));

    bool result = false;
    uint totalPages = 0;

    byte[]? data = null;

    try
    {
        MagickImageCollection magickImages = new MagickImageCollection();

        totalPages = (uint)PdfInfo.Create(@"C:\temp\Libro.pdf").PageCount;

        data = File.ReadAllBytes(@"C:\temp\Libro.pdf");

        Parallel.For(0, (int)totalPages, new ParallelOptions {
MaxDegreeOfParallelism = Environment.ProcessorCount }, index =>
        {
            MagickReadSettings settings = new MagickReadSettings()
            {
                FrameIndex = (uint)index,
                ColorSpace = ColorSpace.sRGB,
                ColorType = ColorType.TrueColor,
                Compression = CompressionMethod.JPEG,
                Density = new Density(150, 150)
            };

            try
            {
                MagickImage image = new MagickImage(data, settings);

                image.Write(System.IO.Path.Combine(@"C:\temp\Pruebas\",
"Page_" + (index + 1) + ".jpg"), MagickFormat.Jpg);

                image.Dispose();
            }
            catch (Exception exc)
            {
            }
            finally
            {

            }
        });

        data = null;
        result = true;
    }
    catch (Exception)
    {
        throw;
    }

    Assert.True(result);

}

El mar, 10 sept 2024 a las 6:54, Dirk Lemstra ***@***.***>)
escribió:

> It sounds like you are missing a using and not disposing an image
> correctly.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dlemstra/Magick.NET/discussions/1710#discussioncomment-10597989>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APJ2PETNGAJ7OMITN7MPD4LZVZ3PNAVCNFSM6AAAAABN25F25SVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJZG44TQOI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Un saludo,

Manuel Quero Delgado

**Comment by manuel-collab** [comment]
Hi Kirk, how can I correctly release image objects with the Dispose Method?

Thanks

El mar, 10 sept 2024 a las 7:14, Manuel Quero Delgado (<
***@***.***>) escribió:

> This is my source code:
>
> [Fact]
> public void GetImageFrameToFrame()
> {
>     ResourceLimits.LimitMemory(new Percentage(5));
>
>     bool result = false;
>     uint totalPages = 0;
>
>     byte[]? data = null;
>
>     try
>     {
>         MagickImageCollection magickImages = new MagickImageCollection();
>
>         totalPages = (uint)PdfInfo.Create(@"C:\temp\Libro.pdf").PageCount;
>
>         data = File.ReadAllBytes(@"C:\temp\Libro.pdf");
>
>         Parallel.For(0, (int)totalPages, new ParallelOptions {
> MaxDegreeOfParallelism = Environment.ProcessorCount }, index =>
>         {
>             MagickReadSettings settings = new MagickReadSettings()
>             {
>                 FrameIndex = (uint)index,
>                 ColorSpace = ColorSpace.sRGB,
>                 ColorType = ColorType.TrueColor,
>                 Compression = CompressionMethod.JPEG,
>                 Density = new Density(150, 150)
>             };
>
>             try
>             {
>                 MagickImage image = new MagickImage(data, settings);
>
>                 image.Write(System.IO.Path.Combine(@"C:\temp\Pruebas\",
> "Page_" + (index + 1) + ".jpg"), MagickFormat.Jpg);
>
>                 image.Dispose();
>             }
>             catch (Exception exc)
>             {
>             }
>             finally
>             {
>
>             }
>         });
>
>         data = null;
>         result = true;
>     }
>     catch (Exception)
>     {
>         throw;
>     }
>
>     Assert.True(result);
>
> }
>
> El mar, 10 sept 2024 a las 6:54, Dirk Lemstra ***@***.***>)
> escribió:
>
>> It sounds like you are missing a using and not disposing an image
>> correctly.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dlemstra/Magick.NET/discussions/1710#discussioncomment-10597989>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/APJ2PETNGAJ7OMITN7MPD4LZVZ3PNAVCNFSM6AAAAABN25F25SVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJZG44TQOI>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Un saludo,
>
> Manuel Quero Delgado
>


-- 
Un saludo,

Manuel Quero Delgado

**Comment by manuel-collab** [comment]
Sorry Dirk!!!

Un saludo,

Manuel Quero Delgado

**Comment by dlemstra** [comment]
I just noticed that your code contains `ResourceLimits.LimitMemory(new Percentage(5));` and that will probably cause ImageMagick to create those temporary files. I don't know why you added this but your performance will improve when you remove this statement.

You also no longer need the `MagickImageCollection`.

And instead of calling `.Dispose()` I would advise you to use a `using` statement instead.

**Comment by manuel-collab** [comment]
Thanks Dirk, but even if I make the changes you have recommended, they
still generate temporary when I run it inside a parallel loop for, I show
you my code:

[Fact]
public void GetImageFrameToFrame()
{
    bool result = false;
    uint totalPages = 0;

    byte[]? data = null;

    data = File.ReadAllBytes(@"C:\temp\Libro.pdf");

    try
    {
        totalPages = PdfInfo.Create(@"C:\temp\Libro.pdf").PageCount;


        Parallel.For(0, totalPages, new ParallelOptions {
MaxDegreeOfParallelism = Environment.ProcessorCount }, index =>
        {
            MagickReadSettings settings = new MagickReadSettings()
            {
                FrameIndex = (uint?)index,
                ColorSpace = ColorSpace.sRGB,
                ColorType = ColorType.TrueColor,
                Compression = CompressionMethod.JPEG,
                Density = new Density(150, 150)
            };

            using (MagickImage image = new MagickImage(data, settings))
            {
                image.Quality = 100;
                image.Write(Path.Combine(@"C:\temp\Pruebas\", "Page_" +
(index + 1) + ".jpg"), MagickFormat.Jpg);
            }
        });

        data = null;
        result = true;
    }
    catch (Exception)
    {
        throw;
    }

    Assert.True(result);
}

El vie, 13 sept 2024 a las 10:55, Dirk Lemstra ***@***.***>)
escribió:

> I just noticed that your code contains ResourceLimits.LimitMemory(new
> Percentage(5)); and that will probably cause ImageMagick to create those
> temporary files. I don't know why you added this but your performance will
> improve when you remove this statement.
>
> You also no longer need the MagickImageCollection.
>
> And instead of calling .Dispose() I would advise you to use a using
> statement instead.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dlemstra/Magick.NET/discussions/1710#discussioncomment-10635365>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APJ2PESR7KEXMX5FYMSEIF3ZWKR67AVCNFSM6AAAAABN25F25SVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANRTGUZTMNI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Un saludo,

Manuel Quero Delgado

**Reply by dlemstra** [reply]
This temp files are used by Ghostscript because when Ghostscript reads a frame it writes it to a temporary output file. But those files should be removed once the images have been read. I thought this was coming from you limiting the memory but I forgot that Ghostscript requires temporary files.

**Comment by manuel-collab** [comment]
Is there any kind of solution to identify the temporary image file name and
delete it after performing the operation? In this case the Dispose method
does not work

Thanks

Manuel Quero Delgado

**Reply by dlemstra** [reply]
Those files should be removed automatically. Are any files left behind?

**Comment by manuel-collab** [comment]
When I run the parallel for loop half of the temp files are not deleted


Un saludo,

Manuel Quero Delgado

**Reply by dlemstra** [reply]
Which version of Magick.NET are you using?

**Comment by manuel-collab** [comment]
13.10.0

Un saludo,

Manuel Quero Delgado

**Reply by dlemstra** [reply]
Please first upgrade to the most recent version and check if the problem still reproduces.

**Comment by manuel-collab** [comment]
The problem still occurs when using the parallel loop


Un saludo,

Manuel Quero Delgado

**Comment by dlemstra** [comment]
On my machine your test completes in 38 seconds and I have no temp files left behind. Are you testing with the same document? And are you also using the latest version of Ghostscript?

One reason for tempfiles being left behind would be an abort of the test run. This kills the process and the temporary files are then not cleaned up.

**Comment by manuel-collab** [comment]
Thanks Dirk, I've followed your advice, I've updated version 14 of
Magick.NET, I've updated the version of Ghostscript to the latest version,
plus I'm passing you the latest implementation of my code, but it still
keeps leaving temporary files.

public static async void Task1(byte[] data, uint totalPages)
{
    try
    {
        object _SyncRoot = new object();



        Parallel.For(0, totalPages, index =>
        {
            MagickReadSettings settings = new MagickReadSettings()
                {
                    FrameIndex = (uint)index,
                    Density = new Density(150, 150)
                };

            using (MagickImage image = new MagickImage(data, settings))
            {
                image.Alpha(AlphaOption.Remove);
                image.Quality = 100;
                image.Write(Path.Combine(@"C:\temp\Pruebas\Libro1",
"Libro_" + (index + 1) + ".png"), MagickFormat.Png);
                image.Dispose();
            }
        });

        DirectoryInfo directory = new
DirectoryInfo(Environment.GetEnvironmentVariable("TEMP",
EnvironmentVariableTarget.User));

        foreach (var file in directory.EnumerateFiles("magick-*"))
        {
            file.Delete();
        }
    }
    catch (Exception exc)
    {
    }
    finally
    {

    }
}

El lun, 16 sept 2024 a las 21:26, Dirk Lemstra ***@***.***>)
escribió:

> On my machine your test completes in 38 seconds and I have no temp files
> left behind. Are you testing with the same document? And are you also using
> the latest version of Ghostscript?
>
> One reason for tempfiles being left behind would be an abort of the test
> run. This kills the process and the temporary files are then not cleaned up.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dlemstra/Magick.NET/discussions/1710#discussioncomment-10663365>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APJ2PETIHP5JBX6KZHHCMTDZW4WGJAVCNFSM6AAAAABN25F25SVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANRWGMZTMNI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Un saludo,

Manuel Quero Delgado

**Comment by manuel-collab** [comment]
private static void Task2(byte[] data, uint totalPages)
        {
            try
            {
                object _SyncRoot = new object();

                Parallel.For(0, totalPages, index =>
                {
                    lock (_SyncRoot)
                    {
                        MagickReadSettings settings = new
MagickReadSettings()
                        {
                            FrameIndex = (uint)index,
                            Density = new Density(150, 150)
                        };

                        using (MagickImage image = new MagickImage(data,
settings))
                        {
                            image.Alpha(AlphaOption.Remove);
                            image.Quality = 100;

image.WriteAsync(Path.Combine(@"C:\temp\Pruebas\Libro2", "Libro2_" + (index
+ 1) + ".png"), MagickFormat.Png);
                            image.Dispose();
                        }
                    }
                });
            }
            catch (Exception exc)
            {
            }
            finally
            {

            }
        }

I don't get temporary files here, but it runs slower

El lun, 16 sept 2024 a las 21:37, Manuel Quero Delgado (<
***@***.***>) escribió:

> Thanks Dirk, I've followed your advice, I've updated version 14 of
> Magick.NET, I've updated the version of Ghostscript to the latest version,
> plus I'm passing you the latest implementation of my code, but it still
> keeps leaving temporary files.
>
> public static async void Task1(byte[] data, uint totalPages)
> {
>     try
>     {
>         object _SyncRoot = new object();
>
>
>
>         Parallel.For(0, totalPages, index =>
>         {
>             MagickReadSettings settings = new MagickReadSettings()
>                 {
>                     FrameIndex = (uint)index,
>                     Density = new Density(150, 150)
>                 };
>
>             using (MagickImage image = new MagickImage(data, settings))
>             {
>                 image.Alpha(AlphaOption.Remove);
>                 image.Quality = 100;
>                 image.Write(Path.Combine(@"C:\temp\Pruebas\Libro1",
> "Libro_" + (index + 1) + ".png"), MagickFormat.Png);
>                 image.Dispose();
>             }
>         });
>
>         DirectoryInfo directory = new
> DirectoryInfo(Environment.GetEnvironmentVariable("TEMP",
> EnvironmentVariableTarget.User));
>
>         foreach (var file in directory.EnumerateFiles("magick-*"))
>         {
>             file.Delete();
>         }
>     }
>     catch (Exception exc)
>     {
>     }
>     finally
>     {
>
>     }
> }
>
> El lun, 16 sept 2024 a las 21:26, Dirk Lemstra ***@***.***>)
> escribió:
>
>> On my machine your test completes in 38 seconds and I have no temp files
>> left behind. Are you testing with the same document? And are you also using
>> the latest version of Ghostscript?
>>
>> One reason for tempfiles being left behind would be an abort of the test
>> run. This kills the process and the temporary files are then not cleaned up.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dlemstra/Magick.NET/discussions/1710#discussioncomment-10663365>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/APJ2PETIHP5JBX6KZHHCMTDZW4WGJAVCNFSM6AAAAABN25F25SVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANRWGMZTMNI>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Un saludo,
>
> Manuel Quero Delgado
>


-- 
Un saludo,

Manuel Quero Delgado

**Reply by dlemstra** [reply]
What happens when you don't put your test file in the `c:\temp` folder? That is the only difference from what I am doing.

p.s. With that `lock` you are just writing a complex for loop.

**Comment by manuel-collab** [comment]
Same problem. It does not delete temporary files. I don't know what else to
do!!

Un saludo,

Manuel Quero Delgado

**Comment by manuel-collab** [comment]
Can you send me a small implementation to see what I might be doing wrong?

Thanks Dirk

El lun, 16 sept 2024 a las 21:52, Manuel Quero Delgado (<
***@***.***>) escribió:

> Same problem. It does not delete temporary files. I don't know what else
> to do!!
>
> Un saludo,
>
> Manuel Quero Delgado
>
>
> El El lun, 16 sept 2024 a las 21:46, Dirk Lemstra <
> ***@***.***> escribió:
>
>> What happens when you don't put your test file in the c:\temp folder?
>> That is the only difference from what I am doing.
>>
>> p.s. With that lock you are just writing a complex for loop.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dlemstra/Magick.NET/discussions/1710#discussioncomment-10663531>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/APJ2PEVMYZ5V4TPXWZW3AV3ZW4YQRAVCNFSM6AAAAABN25F25SVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANRWGM2TGMI>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

-- 
Un saludo,

Manuel Quero Delgado

**Reply by dlemstra** [reply]
I just copy pasted your implementation so I have no idea why those files are left behind. This can only happen when a process is aborted. I cannot reproduce your issue. Can you reproduce the issue in a small application?

**Comment by manuel-collab** [comment]
[Fact]
public void GetImageFrameToFrame()
{
    bool result = false;
    uint totalPages = 0;

    try
    {
        totalPages = PdfInfo.Create(@"C:\Proyectos\Libro.pdf").PageCount;


        Parallel.For(0, totalPages, index =>
        {
            MagickReadSettings settings = new MagickReadSettings()
            {
                FrameIndex = (uint)index,
                Density = new Density(150, 150)
            };

            using (MagickImage image = new
MagickImage(@"C:\Proyectos\Libro.pdf",
settings))
            {
                image.Alpha(AlphaOption.Remove);
                image.Quality = 100;
                image.Write(Path.Combine(@"C:\Proyectos\Pruebas\Libro1",
"Libro_" + (index + 1) + ".png"), MagickFormat.Png);
            }
        });
    }
    catch (Exception)
    {
        throw;
    }

    Assert.True(result);
}

With this example 715 pages and 269 temporary files are not deleted


Un saludo,

Manuel Quero Delgado

**Reply by dlemstra** [reply]
Are you testing with a different file than what you uploaded earlier?

**Comment by manuel-collab** [comment]
No, I always use the same one


Un saludo,

Manuel Quero Delgado

**Comment by dlemstra** [comment]
Your `document.pdf` file only contains `333` pages and not `715`.

**Comment by manuel-collab** [comment]
I also tried it with a 333-page document and it continues to generate
temporary files.


Un saludo,

Manuel Quero Delgado

**Comment by jdmsolarius** [comment]
I cannot reproduce it either. _**I  would focus on a way to very quickly and efficiently delete these temporary files.**_ If you have access to the CLI you can delete them all immediately, if you don't you can Enumerate the Directories in Parallel but be careful how much Parallelism you use how much benefit you get will depend on your hard drive.

1. (Fastest) batch Delete the Files through the CLI if you have access.
```
var command = "del /q C:\\path\\to\\directory\\magick-*";
var startInfo = new ProcessStartInfo("cmd.exe", "/c " + command)
{
    CreateNoWindow = true,
    UseShellExecute = false
};
Process.Start(startInfo);
```
2. If you don't, Enumerate the the Files in Parallel and Delete them (you may have to monitor the number of threads more carefully since this is I/O but the Enumeration can be Parallelized)
```
Parallel.ForEach(Directory.EnumerateFiles(directoryPath, "magick-*"), file =>
{
    File.Delete(file);
});
```
on a three year old m.2 you should be able to delete a Gigabyte of files in no time (how big are these files!?)
